src/AppInstallerRepositoryCore/CompositeSource.cpp

```diff
@@ -574,31 +829,61 @@ namespace AppInstaller::Repository
+            // Search the tracking catalog as it can potentially get better correlations
+            auto trackingCatalog = source.GetTrackingCatalog();
+            SearchResult trackingResult = trackingCatalog.Search(request);
+
+            for (auto&& match : trackingResult.Matches)
             {
-                continue;
-            }
+                // Check for a package already in the result that should have been correlated already.
+                auto packageData = result.CheckForExistingResultFromTrackingPackageMatch(match);
 
-            SearchResult availableResult;
+                // If found existing package in the result, continue
+                if (!packageData)
+                {
+                    continue;
+                }
 
-            try
-            {
-                availableResult = source->Search(request);
-            }
-            catch (...)
-            {
-                LOG_CAUGHT_EXCEPTION();
-                AICLI_LOG(Repo, Warning, << "Failed to search source: " << source->GetDetails().Name);
-                result.AddFailureIfSourceNotPresent({ source->GetDetails().Name, std::current_exception() });
+                // If no package was found that was already in the results, do a correlation lookup with the installed
+                // source to create a new composite package entry if we find any packages there.
+                if (packageData && !packageData->SystemReferenceStrings.empty())
+                {
+                    // Create a search request to run against the installed source
+                    SearchRequest systemReferenceSearch = packageData->CreateInclusionsSearchRequest();
+
+                    // Correlate against installed (allow exceptions out as we own the installed source)
+                    SearchResult installedCrossRef = m_installedSource.Search(systemReferenceSearch);
+
+                    std::shared_ptr<IPackage> installedPackage = GetMatchingPackage(installedCrossRef.Matches,
+                        [&]() {
+                            AICLI_LOG(Repo, Info,
+                                << "Found multiple matches for tracking package [" << match.Package->GetProperty(PackageProperty::Id) <<
+                                "] in source [" << source.GetIdentifier() << "] when searching for [" << systemReferenceSearch.ToString() << "]");
+                        }, [&] {
+                            AICLI_LOG(Repo, Warning, << "  Appropriate installed package could not be determined");
+                        });
+
+                    if (installedPackage && !result.ContainsInstalledPackage(installedPackage.get()))
+                    {
+                        auto compositePackage = std::make_shared<CompositePackage>(
+                            std::move(installedPackage),
+                            GetTrackedPackageFromAvailableSource(result, source, match.Package->GetProperty(PackageProperty::Id)));

```

Comment:
>GetTrackedPackageFromAvailableSource [](http://example.com/codeflow?start=28&length=36)

This might return null, in that case, we prevent the later available package to be associated with this installed package, I guess it'll be fixed when we have support for 1 -> n installed to available mapping in the future

Reply:
Actually, this might cause some upgrade cases to report "no applicable upgrades found" incorrectly. Looks like we may want to fix the 1 -> n mapping sooner.

Reply:
We need to think about the behavior in that case.  We haven't really laid down how the association will work.  In my opinion, we should not attempt to upgrade in the case that the original source no longer contains the package, but another one does.  We should instead inform the user of that fact so that they can decide what to do.  If they did install/upgrade into the other source, then the association would be updated and we would start upgrading from the new location in the future.